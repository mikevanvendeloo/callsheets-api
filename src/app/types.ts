export type ActiveCallSheetTimer = {
  matchStartTime: string;
  activeItem: CallSheetItem;
  nextItem?: CallSheetItem;
  itemTimer: ItemTimer;
}

export type ItemTimer = {
  timerValue: string;
  timerColor: string;
  startTimestamp: string;
  endTimestamp: string;
}

export type CallSheetItem = {
  id: number;
  timeStart: string;
  durationInMinutes: number;
  timeEnd: string;
  title: string;
}

type CallSheet = {
    matchInfo: MatchInfo;
    schedule: Schedule;
}
export default CallSheet

export type MatchInfo = {
    league: string;
    title: string;
    date: string;
    startTime: string;
}
  
export type Schedule = {
    items: ScheduleItem[];
}

export type ScheduleItem = {
    id: number;
    timeStart: string;
    durationInMinutes: number;
    timeEnd: string;
    title: string;
}
  
export type ActiveCallSheet = {
    activeCallSheet: string;
    itemStartTimestamp: string
    activationTimestamp: string;
    activeItem: number;
    callSheet: CallSheet;
}

// export type CallSheetResponse = {
//     matchInfo: MatchInfoResponse;
//     schedule: ScheduleResponse;
// }

// export type ActiveCallSheetResponse = {
//     activeCallSheet: string;
//     itemStartTimestamp: string
//     activationTimestamp: string;
//     activeItem: number;
//     callSheet: CallSheetResponse;
// }


// export type MatchInfoResponse = {
//     league: string;
//     title: string;
//     date: string;
//     startTime: string;
// }
  
// export type ScheduleResponse = {
//     items: ScheduleItemResponse[];
// }

// export type ScheduleItemResponse = {
//     id: number;
//     timeStart: string;
//     durationInMinutes: number;
//     timeEnd: string;
//     title: string;
// }