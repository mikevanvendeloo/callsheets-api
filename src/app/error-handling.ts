import { type Request, type Response, type NextFunction } from 'express'
import type AppError from './api-error'
import { HttpStatusCode } from './api-error'

export const errorLogger = (
  error: Error,
  request: Request,
  response: Response,
  next: NextFunction,
): void => {
  console.log(`error ${error.message}`)
  next(error) // calling next middleware
}

// Error handling Middleware function reads the error message
// and sends back a response in JSON format
export const errorResponder = (
  error: AppError,
  request: Request,
  response: Response,
  next: NextFunction,
): void => {
  response.header('Content-Type', 'application/json')

  const status = error.statusCode ?? HttpStatusCode.BAD_REQUEST
  response.status(status).send(error.message)
}

// Fallback Middleware function for returning
// 404 error for undefined paths
export const invalidPathHandler = (
  request: Request,
  response: Response,
  next: NextFunction,
): void => {
  response.status(404)
  response.send('invalid path')
}
