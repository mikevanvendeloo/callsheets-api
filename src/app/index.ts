import express, { Application } from 'express'
import cors, { CorsOptions } from 'cors'
import Routes from './routes'
import requestLogging from './logging/requestLogging'
import path from 'path'

export const dataDirectory = path.join(__dirname, '../../data')
export const callSheetDirectory = path.join(dataDirectory, 'callsheets')

export default class Server {
  constructor(app: Application) {
    this.config(app)
    new Routes(app)
  }

  private config(app: Application): void {
    const corsOptions: CorsOptions = {
      origin: '*',
      methods: ['GET', 'POST', 'DELETE'],
      exposedHeaders: [
        'Content-Length',
        'origin',
        'X-Requested-With',
        'Content-Type',
        'Accept',
        'Authorization',
      ],
    }

    app.use(cors(corsOptions))
    app.use(express.json())
    app.use(express.urlencoded({ extended: true }))
    app.use(requestLogging)
  }
}
