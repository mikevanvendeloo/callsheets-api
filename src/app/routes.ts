import { Application } from 'express'
import callSheetRoutes from './callsheets/CallSheetRoutes'
import healthRoutes from './health/HealthRoutes'
import liveRoutes from './live/LiveRoutes'

export default class Routes {
  constructor(app: Application) {
    app.use('/health', healthRoutes)
    app.use('/api/live', liveRoutes)
    app.use('/api/callsheets', callSheetRoutes)
  }
}
