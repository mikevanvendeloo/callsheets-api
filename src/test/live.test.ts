import request from 'supertest'
import app from '../server'

describe('Verify live endpoint', () => {
  test('should return the live callsheet', async () => {
    const res = await request(app).get('/api/live')
    expect(res.body).toMatchSnapshot()
  })
})
