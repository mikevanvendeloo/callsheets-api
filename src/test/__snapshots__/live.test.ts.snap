// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify live endpoint should return the live callsheet 1`] = `
{
  "activationTimestamp": "2023-10-29T21:07:15.333+01:00",
  "activeCallSheet": "17-11-2023__resKL__Fortuna-Delta Logistiek - Blauw-Wit.json",
  "activeItem": 1,
  "callsheet": {
    "matchInfo": {
      "date": "17-11-2023",
      "league": "resKL",
      "startTime": "13:30",
      "title": "Fortuna/Delta Logistiek - Blauw/Wit",
    },
    "schedule": {
      "activeItem": null,
      "items": [
        {
          "durationInMinutes": 15,
          "id": 1,
          "timeEnd": "13:10",
          "timeStart": "12:55",
          "title": "Laatste checks",
        },
        {
          "durationInMinutes": 0.5,
          "id": 2,
          "timeEnd": "13:11",
          "timeStart": "13:10",
          "title": "Intro livestream",
        },
        {
          "durationInMinutes": 1,
          "id": 3,
          "timeEnd": "13:12",
          "timeStart": "13:11",
          "title": "Welkom livestream",
        },
        {
          "durationInMinutes": 2,
          "id": 4,
          "timeEnd": "13:14",
          "timeStart": "13:12",
          "title": "Reclames/wedstrijdsponsor",
        },
        {
          "durationInMinutes": 3,
          "id": 5,
          "timeEnd": "13:17",
          "timeStart": "13:14",
          "title": "Interview",
        },
        {
          "durationInMinutes": 3,
          "id": 6,
          "timeEnd": "13:20",
          "timeStart": "13:17",
          "title": "Interview",
        },
        {
          "durationInMinutes": 6,
          "id": 7,
          "timeEnd": "13:26",
          "timeStart": "13:20",
          "title": "Voorbeschouwing, stand, opstelling ploegen in beeld",
        },
        {
          "durationInMinutes": 3,
          "id": 8,
          "timeEnd": "13:29",
          "timeStart": "13:26",
          "title": "Sfeerbeelden vooraf wedstrijd, intro door commentatoren",
        },
        {
          "durationInMinutes": 1,
          "id": 9,
          "timeEnd": "13:30",
          "timeStart": "13:29",
          "title": "Wedstrijdsponsor",
        },
        {
          "durationInMinutes": 0,
          "id": 10,
          "timeEnd": "13:30",
          "timeStart": "13:30",
          "title": "Signaal naar juryvoorzitter/scheidsrechter begin wedstrijd",
        },
        {
          "durationInMinutes": 35,
          "id": 11,
          "timeEnd": "14:05",
          "timeStart": "13:30",
          "title": "1e helft",
        },
        {
          "durationInMinutes": 2,
          "id": 12,
          "timeEnd": "14:07",
          "timeStart": "14:05",
          "title": "Reclames/wedstrijdsponsor",
        },
        {
          "durationInMinutes": 10,
          "id": 13,
          "timeEnd": "14:17",
          "timeStart": "14:07",
          "title": "Analyse + (eventueel replay van acties)",
        },
        {
          "durationInMinutes": 1,
          "id": 14,
          "timeEnd": "14:18",
          "timeStart": "14:17",
          "title": "Reclames/wedstrijdsponsor",
        },
        {
          "durationInMinutes": 3,
          "id": 15,
          "timeEnd": "14:21",
          "timeStart": "14:18",
          "title": "Statistieken + commentaar",
        },
        {
          "durationInMinutes": 35,
          "id": 16,
          "timeEnd": "14:56",
          "timeStart": "14:21",
          "title": "2e helft",
        },
        {
          "durationInMinutes": 2,
          "id": 17,
          "timeEnd": "14:58",
          "timeStart": "14:56",
          "title": "Statistieken + commentaar + sfeerbeelden na wedstrijd.",
        },
        {
          "durationInMinutes": 3,
          "id": 18,
          "timeEnd": "15:01",
          "timeStart": "14:58",
          "title": "Analyse + (eventueel replay van acties)",
        },
        {
          "durationInMinutes": 0.5,
          "id": 19,
          "timeEnd": "15:01",
          "timeStart": "15:01",
          "title": "Reclames/wedstrijdsponsor",
        },
        {
          "durationInMinutes": 3,
          "id": 20,
          "timeEnd": "15:04",
          "timeStart": "15:01",
          "title": "Interview",
        },
        {
          "durationInMinutes": 0.5,
          "id": 21,
          "timeEnd": "15:05",
          "timeStart": "15:04",
          "title": "Reclames/wedstrijdsponsor",
        },
        {
          "durationInMinutes": 3,
          "id": 22,
          "timeEnd": "15:08",
          "timeStart": "15:05",
          "title": "Interview",
        },
        {
          "durationInMinutes": 0.5,
          "id": 23,
          "timeEnd": "15:08",
          "timeStart": "15:08",
          "title": "Reclames/wedstrijdsponsor",
        },
        {
          "durationInMinutes": 2,
          "id": 24,
          "timeEnd": "15:10",
          "timeStart": "15:08",
          "title": "Reactie op interviews, uitslagen, standen en volgende wedstrijd",
        },
        {
          "durationInMinutes": 2,
          "id": 25,
          "timeEnd": "15:12",
          "timeStart": "15:10",
          "title": "Highlights wedstrijd",
        },
        {
          "durationInMinutes": 1,
          "id": 26,
          "timeEnd": "15:13",
          "timeStart": "15:12",
          "title": "Credits crew en outro livestream",
        },
      ],
    },
  },
}
`;
